### PR TITLE
Completely rewritten Add-DbaAgDatabase / Added Test-DbaAvailabilityGroup

### LINKED ISSUE #5917: Add-DbaAgDatabase - Add Wait switch based on Sync status
### Summary of new feature
I am running a synchronous AlwaysOn cluster and when I use Add-DbaAgDatabase the command returns sucessfully even when the secondary databases SynchronizationState is still NotSynchronizing/Initializing. I do not really want to continue with my skript until the database is in sync and a "-wait" switch (cherry on top would be a wait timeout) would make live easier. 

### Proposed technical details (if applicable)

Working on it, if I find a way I will update the request.

### Latest version of dbatools as of writing

1.0.22

### LINKED ISSUE #6420: Add joinonly flag to Add-DbaAgDatabase
### Summary of new feature
Add a joinonly flag to be used if the database is already restored to the secondary replica.

### Proposed technical details (if applicable)
Same functionality as join only in SSMS GUI

Add-SqlAvailabilityDatabase -Path "SQLSERVER:\sql\sql01\inst01\AvailabilityGroups\AG01" -database db1

### Latest version of dbatools as of writing
1.0.97

### LINKED ISSUE #6464: Add -ReuseSourceFolderStructure parameter to ADD-DBAAgDatabase
### Summary of new feature

Currently when using manual seeding with Add-DBAAgDatabase you can only use the destination default folders for data and log files.  Adding this flag, which is in Restore-DbaDatabase, would allow Add-DBAAgDatabase to have the same functionality as Restore-DbaDatabase

### Proposed technical details (if applicable)

In the event that there are multiple locations for data and/or log files currently you can't use Add-DBAAgDatabase with manual seeding as all files will go the default locations on the destination server.  By adding this parameter you can match the source file folders, just like autoseed does.  The change in the cmdlet would be to check if teh parameter was included and then add it to the Restore-DbaDatabase call.  

### Latest version of dbatools as of writing

1.0.104

### PR DESCRIPTION
**Pull request by andreasjordan**
## Type of Change
 - [ ] Bug fix (non-breaking change, fixes # )
 - [x] New feature (non-breaking change, adds functionality, fixes #6420, fixes #6464, fixes #5917 )
 - [ ] Breaking change (effects multiple commands or functionality, fixes # )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 
### Purpose
Adding more functionality.

### CODE REVIEW

File: functions/Add-DbaAgDatabase.ps1 [hunk 1]
```diff
@@ -71,7 +94,7 @@ function Add-DbaAgDatabase {
 
     .NOTES
         Tags: AvailabilityGroup, HA, AG
-        Author: Chrissy LeMaire (@cl), netnerds.net
+        Author: Chrissy LeMaire (@cl), netnerds.net / Andreas Jordan (@JordanOrdix), ordix.de
```
Comment (wsmelton) [inline functions/Add-DbaAgDatabase.ps1:97]:
```suggestion
        Author: Chrissy LeMaire (@cl), netnerds.net | Andreas Jordan (@JordanOrdix), ordix.de
```

Has to be pipe symbol for our doc generation to parse it correctly

File: functions/Add-DbaAgDatabase.ps1 [hunk 2]
```diff
@@ -101,174 +124,367 @@ function Add-DbaAgDatabase {
     #>
     [CmdletBinding(SupportsShouldProcess, ConfirmImpact = 'Low')]
     param (
-        [DbaInstanceParameter[]]$SqlInstance,
+        [Parameter(ParameterSetName = 'NonPipeline', Mandatory = $true, Position = 0)]
+        [DbaInstanceParameter]$SqlInstance,
+        [Parameter(ParameterSetName = 'NonPipeline')]
         [PSCredential]$SqlCredential,
-        [parameter(Mandatory)]
+        [Parameter(ParameterSetName = 'NonPipeline', Mandatory = $true)][Parameter(ParameterSetName = 'Pipeline', Mandatory = $true, Position = 0)]
```
Comment (wsmelton) [inline functions/Add-DbaAgDatabase.ps1:131]:
```suggestion
        [Parameter(ParameterSetName = 'NonPipeline', Mandatory = $true)]
        [Parameter(ParameterSetName = 'Pipeline', Mandatory = $true, Position = 0)]
```
For easier reading

File: functions/Add-DbaAgDatabase.ps1 (hunk 1 above)
Comment (andreasjordan) [reply functions/Add-DbaAgDatabase.ps1:97]: Thanks.

File: functions/Add-DbaAgDatabase.ps1 (hunk 2 above)
Comment (andreasjordan) [reply functions/Add-DbaAgDatabase.ps1:131]: Ok, will do so.

**Comment by potatoqualitee** [comment]
how're we looking @wsmelton ? Thank you both for your work on this 🌈

**Comment by wsmelton** [comment]
LGTM

_Note I don't have environment anymore to do integration testing for AGs but logically it looks right._

**Comment by potatoqualitee** [comment]
I have not tested but code looks reasonable to Shawn, and I see a passed integration test [on appveyor](https://ci.appveyor.com/project/sqlcollaborative/dbatools/builds/39481239/job/ytn10l7v60y0dr44).

I'll add new command to manifest and release today, thank you!

**Pull request merged**